api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -744,54 +750,69 @@ private FirecloudWorkspaceId generateFirecloudWorkspaceId(String namespace, Stri
-    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
-    fileDetail.setLastModifiedTime(now.getTime());
+    FileDetail fileDetail = notebookCloneOperation(workspace, workspaceName, rename.getName(), newName);
+    notebookDeleteOperation(workspace, workspaceName, rename.getName());
     return ResponseEntity.ok(fileDetail);
   }
 
   @Override
   public ResponseEntity<FileDetail> cloneNotebook(String workspace, String workspaceName, String notebookName) {
-    String bucket = fireCloudService.getWorkspace(workspace, workspaceName)
-      .getWorkspace()
-      .getBucketName();
-    String origPath = NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + notebookName;
     String newName = notebookName.replaceAll("\\.ipynb", " ") + "Clone.ipynb";
-    String newPath = NOTEBOOKS_WORKSPACE_DIRECTORY + "/" + newName;
-    BlobId blobId = BlobId.of(bucket, origPath);
-    cloudStorageService.copyBlob(blobId, BlobId.of(bucket, newPath));
+    FileDetail fileDetail = notebookCloneOperation(workspace, workspaceName, notebookName, newName);
+    return ResponseEntity.ok(fileDetail);
+  }
+
+  @Override
+  public ResponseEntity<EmptyResponse> deleteNotebook(String workspace, String workspaceName, String notebookName) {
+    notebookDeleteOperation(workspace, workspaceName, notebookName);
+    return ResponseEntity.ok(new EmptyResponse());
+  }
+
+  private FileDetail notebookCloneOperation(String workspace, String workspaceName, String notebookName, String newName) {
+    Map<String, Object> map = notebookOpSetup(workspace, workspaceName, notebookName, newName);
     FileDetail fileDetail = new FileDetail();
+    cloudStorageService.copyBlob((BlobId) map.get("blobId"), (BlobId) map.get("newBlobId"));
     fileDetail.setName(newName);
-    fileDetail.setPath("gs://" + bucket + "/" + newPath);
+    fileDetail.setPath(map.get("fullPath").toString());
     Timestamp now = new Timestamp(clock.instant().toEpochMilli());
     fileDetail.setLastModifiedTime(now.getTime());
-    return ResponseEntity.ok(fileDetail);
+    userRecentResourceService.updateNotebookEntry(objToLong(map.get("workspaceId")), objToLong(map.get("userId")), map.get("fullPath").toString(), now);
+    return fileDetail;
   }
 
-  @Override
-  public ResponseEntity<EmptyResponse> deleteNotebook(String workspace, String workspaceName, String notebookName) {
+  private void notebookDeleteOperation(String workspace, String workspaceName, String notebookName) {
+    Map<String, Object> map = notebookOpSetup(workspace, workspaceName, notebookName, "");
+    cloudStorageService.deleteBlob((BlobId) map.get("blobId"));
+    userRecentResourceService.deleteNotebookEntry(objToLong(map.get("workspaceId")), objToLong(map.get("userId")), map.get("fullPath").toString());
+  }
+
+  private long objToLong(Object object) {
+    return Long.parseLong(String.valueOf(object));
+  }
+
+  private Map<String, Object> notebookOpSetup(String workspace, String workspaceName, String notebookName, String newName) {
```

Comment:
I can see what you're trying to do here with `Map<String, Object>` but that leads to extra code, ie. `objToLong` that you then need to de-convolute what's in the map. In my opinion, it would be cleaner to just make a new DTO class that holds the variables you want with the types that you want so you don't have to munge-unmunge data.